### DISCUSSION TITLE
Emit With "to" Definition Sometimes Doesn't Work

### DISCUSSION THREAD

**Discussion by yakirs57** [miguelgrinberg/python-socketio#1181] [Q&A]
Hi!

I am experiencing an issue in emitting calls from the server to the client.
Just to clarify, when I am sending an emit from the client to the server and then, from the event handler on the server, I send an emit back to the client it always arrives at the client.
**BUT**
When I am trying to emit to the client outside of a client's emit event handler, the emit doesn't arrive at the client every time.
After exploring the situation a little further, I've figured out that the session of the socket connection is not available to all the workers of the Flask application. When the emit works it is only when the socket connection is available to the worker that manages the request.

Is there a way to solve it?
I am running a Flask application with over 150 workers and preload settings.

I expect that the socket session will be available to all workers of the application for the purpose of sending an emit from a different worker of the server to the client.

Please Advise.

**Comment by miguelgrinberg** [comment]
Please do not write issues on this package when you are just asking a question. The discussion board is for question, the issues board is for bugs.

Regarding your question, have you read the documentation? [Start here](https://python-socketio.readthedocs.io/en/latest/server.html#using-a-message-queue) if you haven't.

**Reply by yakirs57** [reply]
I understand. Thank you for the clarification on conduct.

Regarding the client manager, I will test it, but it seems like the exact solution I need. Does it affect performance or require any changes in how I access the socket? It doesn't mention anything in the documentation but just to make sure.

**Reply by miguelgrinberg** [reply]
It will affect performance because when you want to emit to some client a search has to be conducted to see which of the workers has the socket for that client. But this is all handled transparently, there are no changes to your code required, beyond configuring the client manager.

**Reply by yakirs57** [reply]
In case you emit without defining "to" in the kwargs. Meaning it sends to the socket connected to the active request.
Does it still have the same effect on the performance?

**Reply by miguelgrinberg** [reply]
I don't understand your question. Are you using python-socketio or Flask-SocketIO? In this package (python-socketio) there is no concept of "active request", that doesn't mean anything. When you emit, you have to specify the "to", and if you don't then the emit is a broadcast that goes out to everybody.

**Reply by yakirs57** [reply]
I am using Flask-SocketIO. Let me clarify my question.
You mentioned that using the client manager will affect the performance because when I emit "a search has to be conducted to see which of the workers has the socket for that client".
Now I am asking about a situation: I have a socket endpoint on the server, that received a message from one of the connected clients, and I want to send a message back to that same client. Now, in that situation, it is 100% certain that it uses the same Worker as the socket connection. Would it still go through the search you have mentioned and still affect the performance? Or does it first search locally and then go to search in the rest of the Workers?
Regarding not specifying the "to", it's off-topic, but I meant that when you are on a socket endpoint that just received a message from a connected client, and now you want to send an emit to the client you received the message from, you don't need to specify the "to" because when the to is None and the room is None it will default to the request session id (sid / flask.request.sid).

**Reply by miguelgrinberg** [reply]
If you want to avoid this search I was talking about, you can pass the `ignore_queue=True` option.
